lib/pos_mempool.go

```diff
@@ -0,0 +1,391 @@
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem validating transaction sanity")
+	}
+
+	// Construct the MempoolTx from the MsgDeSoTxn.
+	mempoolTx, err := NewMempoolTx(txn, uint64(latestBlockHeight))
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem constructing MempoolTx")
+	}
+
+	// Check transaction signature
+	if _, err := dmp.latestBlockView.VerifySignature(txn, latestBlockHeight); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Signature validation failed")
+	}
+
+	// Add the transaction to the mempool.
+	if err := dmp.AddTransaction(mempoolTx); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding transaction to mempool")
+	}
+	return nil
+}
+
+// AddTransaction is the main function for adding a new transaction to the mempool.
+func (dmp *PosMempool) AddTransaction(txn *MempoolTx) error {
+	if !dmp.IsRunning() {
+		return errors.Wrapf(MempoolErrorNotRunning, "PosMempool.AddTransaction: ")
+	}
+
+	dmp.Lock()
+	defer dmp.Unlock()
+
+	// Add the transaction to the mempool and then prune if needed.
+	if err := dmp.addTransactionNoLock(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding transaction to mempool")
+	}
+
+	if err := dmp.pruneNoLock(); err != nil {
+		glog.Errorf("PosMempool.AddTransaction: Problem pruning mempool: %v", err)
+	}
+
+	return nil
+}
+
+func (dmp *PosMempool) addTransactionNoLock(txn *MempoolTx) error {
+	userPk := NewPublicKey(txn.Tx.PublicKey)
+	txnFee := txn.Tx.TxnFeeNanos
+	latestBlockHeight := dmp.latestBlockNode.Height
+
+	// Validate that the user has enough balance to cover the transaction fees.
+	spendableBalanceNanos, err := dmp.latestBlockView.GetSpendableDeSoBalanceNanosForPublicKey(userPk.ToBytes(), latestBlockHeight)
+	if err != nil {
+		return errors.Wrapf(err, "CheckBalanceIncrease: Problem getting spendable balance")
+	}
+	if err := dmp.ledger.CanIncreaseBalance(*userPk, txnFee, spendableBalanceNanos); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem checking balance increase for transaction with"+
+			"hash %v, fee %v", txn.Tx.Hash(), txnFee)
+	}
+
+	// If we get here, it means that the transaction's sender has enough balance to cover transaction fees. We can now
+	// add the transaction to mempool.
+	if err := dmp.txnRegister.AddTransaction(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding txn to register")
+	}
+
+	// We update the reserved balance to include the newly added transaction's fee.
+	dmp.ledger.IncreaseBalance(*userPk, txnFee)
+
+	// Emit an event for the newly added transaction.
+	if dmp.emitEvents {
+		event := &MempoolEvent{
+			Txn:  txn,
+			Type: MempoolEventAdd,
+		}
+		dmp.persister.EnqueueEvent(event)
+	}
+
+	return nil
+}
+
+// RemoveTransaction is the main function for removing a transaction from the mempool.
+func (dmp *PosMempool) RemoveTransaction(txn *MempoolTx) error {
```

Comment:
Would it make sense to update the input parameter here to be the txn hash? In practice, an external caller likely won't have the entire MempoolTx available on-hand. In order to remove a txn, they would first have to fetch up the txn from the mempool by its hash, and pass the txn here. We may as well simplify it for them and have them pass in the txn hash here anyway
```suggestion
func (dmp *PosMempool) RemoveTransaction(txnHash *BlockHash) error {
```